secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/dcr/request/RequestAndSsaSignatureValidationFilter.java

```diff
@@ -0,0 +1,319 @@
+        Reject.ifNull(registrationRequestObjectFromJwtSupplier, "RegistrationRequestObjectFromJwtSupplier " +
+                "must be provided");
+        Reject.ifNull(supportedSigningAlgorithms, "supportedSigningAlgorithms must be provided");
+        Reject.ifNull(registrationRequestJwtValidator, "registrationRequestJwtValidator must be provided");
+        this.registrationRequestObjectFromJwtSupplier = registrationRequestObjectFromJwtSupplier;
+        this.supportedSigningAlgorithms = supportedSigningAlgorithms;
+        this.ssaValidator = ssaValidator;
+        this.registrationRequestJwtValidator = registrationRequestJwtValidator;
+        this.dcrUtils = dcrUtils;
+        log.debug("RequestAndSsaSignatureValidationFilter constructed");
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        final String fapiInteractionId = FAPIUtils.getFapiInteractionIdForDisplay(context);
+        log.debug("({}) filtering - filtering out invalid signatures of registration request and ssa jwts",
+                fapiInteractionId);
+        if (!VALIDATABLE_HTTP_REQUEST_METHODS.contains(request.getMethod())) {
+            return next.handle(context, request);
+        }
+
+        try {
+            SignedJwt registrationRequestJwt = getRegistrationRequestObjectOrThrow(fapiInteractionId, context, request);
+            checkJwtSigningAlgorithmIsValid(fapiInteractionId, registrationRequestJwt, supportedSigningAlgorithms);
+            final JwtClaimsSet registrationRequestJwtClaimsSet = registrationRequestJwt.getClaimsSet();
+
+            final String ssaJwtString = getSsaB64EncodedJwtString(fapiInteractionId, registrationRequestJwtClaimsSet);
+            log.debug("({}) ssa from registration request jwt is {}", fapiInteractionId, ssaJwtString);
+            SignedJwt ssaSignedJwt = getSoftwareStatementAssertionSignedJwt(fapiInteractionId, ssaJwtString, supportedSigningAlgorithms);
+            final JwtClaimsSet ssaClaimsSet = ssaSignedJwt.getClaimsSet();
+
+            Promise<Response, DCRSignatureValidationException> ssaValidationPromise =
+                    ssaValidator.validateSoftwareStatementAssertionSignature(fapiInteractionId, ssaSignedJwt);
+            return ssaValidationPromise.thenAsync( response ->
+                registrationRequestJwtValidator.validateRegistrationRequestJwtSignature(
+                            fapiInteractionId, ssaClaimsSet, registrationRequestJwt)
+                            .thenAsync(regRequestValidationResponse -> {
+                        if (regRequestValidationResponse.getStatus() != Status.OK) {
+                            return Promises.newResultPromise(response);
+                        }
+                        return next.handle(context, request);
+                    }, ex -> {
+                        return Promises.newResultPromise(new Response(Status.BAD_REQUEST).setEntity(getJsonResponseBody(ex)));
```

Comment:
When does this exception handler get called?

Also can you take a look at reformatting this async chain please? There are 2 checked exception handlers at the same indentation, making it hard to tell which async call they relate to.

It may be more readable to extract methods that have a sig which matches the org.forgerock.util.AsyncFunction, and then use a method reference, or have a method that returns a org.forgerock.util.AsyncFunction object. Then use this pattern when the chaining starts to get complex

Reply:
Ah, I meant to test all that and forgot. Will review and rationalise all that lump of handling.